### PR TITLE
Add a PID controller for basic tuning / search / optimization

### LINKED ISSUE #17271: Add simple parametric search capability using PID control
## Reason
Be able to adjust pump power to make temperature delta match in a heated channel during a pseudo transient

## Design
A simple PID controller would suffice, leveraging the control system.
Since this is for pseudo transients, I would leave the convergence check at the simulation level, instead of looking at how well the control is achieving the target 

## Impact
New capability in framework, though redundant with more specific optimization applications
Widely applicable for simple 1D parametric search

### PR DESCRIPTION
**Pull request by GiudGiud**
Closes #17271

### CODE REVIEW

File: framework/include/controls/PIDControl.h [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+#pragma once
+
+// MOOSE includes
+#include "Control.h"
+
+// Forward declarations
+class PIDControl;
+
+template <>
+InputParameters validParams<PIDControl>();
```
Comment (GiudGiud) [inline framework/include/controls/PIDControl.h:19]: I dont need this anymore do I

File: framework/src/controls/PIDControl.C [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+// MOOSE includes
+#include "PIDControl.h"
+
+registerMooseObject("MooseApp", PIDControl);
+
+defineLegacyParams(PIDControl);
```
Comment (GiudGiud) [inline framework/src/controls/PIDControl.C:15]: same?

File: framework/include/controls/PIDControl.h (hunk 1 above)
Comment (loganharbour) [reply framework/include/controls/PIDControl.h:19]: Correct - for new objects we don't setup the legacy parameters. You can also remove the forward declarations above on 15-16; these were only needed with the old legacy params

File: framework/src/controls/PIDControl.C (hunk 2 above)
Comment (loganharbour) [reply framework/src/controls/PIDControl.C:15]: Yep remove

**Comment by lindsayad** [comment]
cool stuff! I wish my thermostat had the D part of PID 😄

**Comment by GiudGiud** [comment]
Is there a cleaner way to handle resets during failed Picard iterations? 
Maybe using the restart system?

File: framework/include/controls/PIDTransientControl.h [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  /// The coefficient multiplying the integral of the error
+  const Real _Kint;
+  /// The coefficient multiplying the error
+  const Real _Kpro;
+  /// The coefficient multiplying the derivative of the error
+  const Real _Kder;
+  /// The time to start the PID controller on
+  const Real _start_time;
+  /// The time to stop using the PID controller on
+  const Real _stop_time;
```
Comment (andrsd) [inline framework/include/controls/PIDTransientControl.h:48]: Should not enabling/disabling of the object be done via the `TimePeriod` class? Or we are not able to enable/disable `Control`-derived objects? Just asking...

File: framework/src/controls/PIDTransientControl.C [hunk 4]
```diff
@@ -0,0 +1,148 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+// MOOSE includes
```
Comment (andrsd) [inline framework/src/controls/PIDTransientControl.C:10]: Delete this line.

File: framework/src/controls/PIDTransientControl.C [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+registerMooseObject("MooseApp", PIDTransientControl);
+
+InputParameters
+PIDTransientControl::validParams()
+{
+  InputParameters params = Control::validParams();
+  params.addClassDescription(
+      "Sets the value of a 'Real' input parameter (or postprocessor) based on a Proportional "
+      "Integral "
+      "Derivative control of a postprocessor to match a target a target value.");
```
Comment (andrsd) [inline framework/src/controls/PIDTransientControl.C:24]: Reflow this. Looks like a bad `clang-format` side effect of formatting already broken up strings.

File: framework/src/controls/PIDTransientControl.C [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+    _Kpro(getParam<Real>("K_proportional")),
+    _Kder(getParam<Real>("K_derivative")),
+    _start_time(getParam<Real>("start_time")),
+    _stop_time(getParam<Real>("stop_time")),
+    _reset_every_timestep(getParam<bool>("reset_every_timestep")),
+    _reset_integral_windup(getParam<bool>("reset_integral_windup"))
+{
+  _integral = 0;
+
+  // To keep the previous values if a multiapp coupling iteration fails
```
Comment (andrsd) [inline framework/src/controls/PIDTransientControl.C:70]: This comment is confusing. How does this keep any old values? This sets old values to `0`.

And why are not these values set in the initializer list?

File: framework/src/controls/PIDTransientControl.C [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+    }
+
+    // Compute the delta between the current value of the postprocessor and the desired value
+    Real delta = _current - _target.value(_t, dummy);
+
+    // If desired, reset integral of the error if the error crosses zero
+    if (_reset_integral_windup && delta * _old_delta < 0)
+    {
+      _integral = 0;
+    }
```
Comment (andrsd) [inline framework/src/controls/PIDTransientControl.C:130]: You may want to remove the `{` `}`. Looks like you do that in other places already...

File: framework/src/controls/PIDTransientControl.C [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+
+    // Compute the three error terms and add them to the controlled value
+    _integral += delta * _dt;
+    value += _Kint * _integral + _Kpro * delta;
+    if (_dt > 0)
+      value += _Kder * delta / _dt;
+
+    if (isParamValid("parameter"))
+      // Set the new value using the Control system
```
Comment (andrsd) [inline framework/src/controls/PIDTransientControl.C:139]: cpt. Obvious

File: framework/src/controls/PIDTransientControl.C [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+      setControllableValue<Real>("parameter", value);
+    else
+      // Set the new postprocessor value using the FE Problem
```
Comment (andrsd) [inline framework/src/controls/PIDTransientControl.C:142]: cpt. Obvious

File: framework/src/controls/PIDTransientControl.C [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+      _integral_old = _integral;
+      _value_old = value;
+      _t_step_old = _t_step;
+      _old_delta = 0;
+    }
+
+    // If there were coupling/Picard iterations during the transient and they failed,
+    // we need to reset the controlled value and the error integral to their initial value at the
+    // beginning of the coupling process
+    if (dynamic_cast<Transient *>(_app.getExecutioner())->picardSolve().numPicardIts() == 1)
```
Comment (andrsd) [inline framework/src/controls/PIDTransientControl.C:117]: 1. `picardSolve()` is in `Executioner`, so why the `down_cast` to `Transient`?
2. if you want to make sure people use your `Control` object only with `Transient` executioner, you should check that in the c-tor, maybe?

**Review by andrsd — CHANGES_REQUESTED** [review]

File: framework/src/controls/PIDTransientControl.C (hunk 6 above)
Comment (GiudGiud) [reply framework/src/controls/PIDTransientControl.C:70]: true this comment fits better in the header

File: framework/src/controls/PIDTransientControl.C (hunk 10 above)
Comment (GiudGiud) [reply framework/src/controls/PIDTransientControl.C:117]: I think I was being over zealous with the NumPicardIts pp not being supported for steady executioner. I am not sure why though.

File: framework/include/controls/PIDTransientControl.h (hunk 3 above)
Comment (GiudGiud) [reply framework/include/controls/PIDTransientControl.h:48]: I get this:
*** ERROR ***
The following error occurred in the object "diff", of type "TimePeriod".

The desired parameter 'Controls/integral_value/enable' was not located for the 'diff' object, it either does not exist or has not been declared as controllable.

Comment (andrsd) [reply framework/include/controls/PIDTransientControl.h:45]: Ok, thanks for try this. This may not be possible on purpose. I just wanted to be sure.

**Comment by andrsd** [comment]
One question for you: Do you want this code to work with `Steady` executioner? Based on the new patch, it seems like you do. Just making sure that if somebody runs your code with it, they won't get some garbage back.

**Comment by GiudGiud** [comment]
PID doesnt make too much sense for a Steady solve. My concern is the Picard iterations. 

Let me check that it works.

**Comment by andrsd** [comment]
>My concern is the Picard iterations.

I understand that. My point is that if somebody runs your code with `Steady`, you probably want to catch that and tell them not do do that :-)

**Comment by GiudGiud** [comment]
Ok it does not return garbage. My example just converges without the PID doing its thing. The residual is 0 after the first solve. I added a picard_min_its parameter and let the PID act and it got the right result for the control. 

I don't think a user should expect a PIDTransientControl to work with a Steady executioner anyway. I think using a min number of Picard iterations to 'make it do its thing' is the right solution there. 

I am adding the minimum iterations parameter in #17593

**Comment by GiudGiud** [comment]
ok adding an error!

**Review by andrsd — APPROVED** [review]

**Pull request merged**